Fix major bug in v0.3.1: negative counts

sjfleming opened the pull request:
Closes #306 
Closes #348 

Something about the update from v0.3.0 to v0.3.1 induced a bug (I'm not yet sure of the scope, but potentially affecting all datasets) where negative counts could be produced in the output.  It is assumed that these outputs are totally compromised.  v0.3.1 has already been redacted, but the master branch is still compromised.

I think the bug probably lies in the updated posterior generation code, which was the target of memory-saving updates in v0.3.1.

Expand tests covering that code.

- unit tests for posterior generation
- integration tests ensuring no negative count matrix entries in the output (not explicitly tested currently)

sjfleming commented:
After running simulated data through v0.3.0 and the master branch, I see that the posterior generated in each case is identical.

So despite what I initially thought, it seems like the bug must have been introduced in `estimation.py`, which is quite well-covered with tests...

(It is not a CPU / GPU issue, as a run without `--cuda` basically agrees with a GPU run.)

sjfleming commented:
It came down to a small typo / miscalculation about what the maximum possible number of noise counts can be.  I had put `np.uint8` in `estimation.py` line 241 [here](https://github.com/broadinstitute/CellBender/commit/3a4dc8de7157f78e71c1689a536392a3e02317e8), but `np.uint8` is not large enough.  This was probably causing integer overflows that affected the noise estimation process in unexpected ways (which admittedly I have not completely understood yet).

Changing to `np.uint8` seems to fix the problem.

**Before**
![image](https://github.com/broadinstitute/CellBender/assets/10214815/1bea684f-3fd4-4e4d-b5c5-2c9f7d3ec8a3)
And there are negative count matrix entries in the output.

**After**
![image](https://github.com/broadinstitute/CellBender/assets/10214815/ba3ecdc1-0b08-4218-87a9-26f1ee6ce6a9)
And there are no longer negative count matrix entries in the output.

mbabadi commented on cellbender/remove_background/tests/test_estimation.py:
```diff
@@ -92,11 +93,12 @@ def test_mean_massive_m(log_prob_coo):
     new_shape = (coo.shape[0] + greater_than_max_int32, coo.shape[1])
     new_coo = sp.coo_matrix((coo.data, (new_row, coo.col)),
                             shape=new_shape)
+    print(new_coo)
```
Either remove the print statement, or convert it to something more informative-- e.g. old matrix shape = ..., new matrix shape = ...

mbabadi reviewed (changes requested):
Nice job expanding the tests!  Just a small comment to address.  Also, while you're at it, you may want to search the rest of the codebase for uint8 and other explicit dtypes and refactor accordingly.

sjfleming commented:
Alright @mbabadi , what do you think now?  I added a warning to the README, and I added a command-line argument to enable people to re-run their old checkpointed v0.3.1 runs to produce a valid output.

mbabadi reviewed (approved):
Looks good!
